File: backend-cpp/src/scheduler/Jobs.cpp
#include "Jobs.h"

#include "../config/Config.h"
#include "../db/ConnectionPool.h"
#include "../fmi/FmiClient.h"
#include "../fmi/WfsParser.h"
#include "../services/FetchLogService.h"
#include "../services/ForecastService.h"
#include "../services/ObservationService.h"
#include "../services/StationService.h"

#include <chrono>
#include <ctime>
#include <drogon/drogon.h>
#include <sstream>

namespace weather::scheduler {

// ─── Time helpers ──────────────────────────────────────────────────────────

static std::string fmtIso(std::chrono::system_clock::time_point tp) {
    time_t t = std::chrono::system_clock::to_time_t(tp);
    struct tm utc{};
    gmtime_r(&t, &utc);
    char buf[32];
    strftime(buf, sizeof(buf), "%Y-%m-%dT%H:%M:%SZ", &utc);
    return buf;
}

// ─── fetchObservations ─────────────────────────────────────────────────────

void fetchObservations() {
    auto now       = std::chrono::system_clock::now();
    auto startTime = now - std::chrono::minutes(20);
    auto endTime   = now;

    std::string startIso = fmtIso(startTime);
    std::string endIso   = fmtIso(endTime);

    auto& pool   = getPool();
    auto& config = getConfig();

    std::string paramsJson =
        "{\"bbox\":\"" + config.fmi_bbox + "\","
        "\"start\":\"" + startIso + "\","
        "\"end\":\""   + endIso   + "\"}";

    long long logId = 0;
    try {
        logId = services::insertFetchLog(pool, "observations", paramsJson);
    } catch (const std::exception& e) {
        LOG_ERROR << "fetchObservations: insertFetchLog failed: " << e.what();
        logId = 0;
    }

    int count = 0;
    std::string errorMsg;

    try {
        fmi::FmiClient client;
        auto result = client.fetchObservations(startIso, endIso, config.fmi_bbox);
        if (!result.success) throw std::runtime_error(result.error);

        WfsParser parser;
        auto parsed = parser.parseObservations(result.xml);

        services::upsertStations(pool, parsed.stations);
        services::upsertObservations(pool, parsed.observations);
        count = static_cast<int>(parsed.observations.size());
        LOG_INFO << "fetchObservations: " << count << " records";
    } catch (const std::exception& e) {
        errorMsg = e.what();
        LOG_ERROR << "fetchObservations failed: " << errorMsg;
    }

    if (logId > 0) {
        try {
            std::optional<std::string> err;
            if (!errorMsg.empty()) err = errorMsg;
            services::updateFetchLog(pool, logId,
                errorMsg.empty() ? "success" : "error",
                count, err);
        } catch (...) {}
    }
}

// ─── fetchForecasts ────────────────────────────────────────────────────────

void fetchForecasts() {
    auto& pool   = getPool();
    auto& config = getConfig();

    auto now    = std::chrono::system_clock::now();
    std::string fetchedAt = fmtIso(now);

    for (const auto& place : config.default_places) {
        long long logId = 0;
        std::string paramsJson = "{\"place\":\"" + place + "\"}";
        try {
            logId = services::insertFetchLog(pool, "forecast_harmonie", paramsJson);
        } catch (...) {}

        int count = 0;
        std::string errorMsg;

        try {
            fmi::FmiClient client;
            auto result = client.fetchForecastHarmonie(place);
            if (!result.success) throw std::runtime_error(result.error);

            WfsParser parser;
            auto parsed = parser.parseForecasts(result.xml, place, fetchedAt);

            if (parsed.station) {
                std::vector<services::StationMeta> sm = {*parsed.station};
                services::upsertStations(pool, sm);
            }

            services::upsertForecasts(pool, parsed.forecasts);
            count = static_cast<int>(parsed.forecasts.size());
            LOG_INFO << "fetchForecasts[" << place << "]: " << count << " records";
        } catch (const std::exception& e) {
            errorMsg = e.what();
            LOG_ERROR << "fetchForecasts[" << place << "] failed: " << errorMsg;
        }

        if (logId > 0) {
            try {
                std::optional<std::string> err;
                if (!errorMsg.empty()) err = errorMsg;
                services::updateFetchLog(pool, logId,
                    errorMsg.empty() ? "success" : "error",
                    count, err);
            } catch (...) {}
        }
    }
}

} // namespace weather::scheduler

File: backend/app/scheduler/jobs.py
import logging
from datetime import datetime, timedelta, timezone

from ..config import settings
from ..database import AsyncSessionLocal
from ..fmi import client as fmi_client
from ..fmi import parsers
from ..models.fetch_log import FetchLog
from ..services import forecast_service, observation_service, station_service

logger = logging.getLogger(__name__)


async def fetch_observations_job() -> None:
    """Fetch the latest 20-minute window of observations from FMI."""
    now = datetime.now(timezone.utc)
    starttime = now - timedelta(minutes=20)
    endtime = now

    async with AsyncSessionLocal() as db:
        log = FetchLog(
            job_type="observations",
            status="running",
            query_params={
                "bbox": settings.fmi_bbox,
                "start": starttime.isoformat(),
                "end": endtime.isoformat(),
            },
        )
        db.add(log)
        await db.commit()
        await db.refresh(log)

        try:
            result = await fmi_client.fetch_observations(starttime, endtime, settings.fmi_bbox)

            station_dicts = parsers.parse_stations(result.location_metadata)
            await station_service.upsert_stations(db, station_dicts)

            obs_rows = parsers.parse_observations(result.data, result.location_metadata)
            count = await observation_service.upsert_observations(db, obs_rows)

            log.status = "success"
            log.records_fetched = count
            log.finished_at = datetime.now(timezone.utc)
            await db.commit()
            logger.info(f"Observations fetched: {count} records")
        except Exception as exc:
            logger.error(f"Observation fetch failed: {exc}")
            log.status = "error"
            log.error_message = str(exc)
            log.finished_at = datetime.now(timezone.utc)
            await db.commit()


async def fetch_harmonie_job() -> None:
    """Fetch Harmonie forecasts for all configured places."""
    now = datetime.now(timezone.utc)

    for place in settings.places_list:
        async with AsyncSessionLocal() as db:
            log = FetchLog(
                job_type="forecast_harmonie",
                status="running",
                query_params={"place": place},
            )
            db.add(log)
            await db.commit()
            await db.refresh(log)

            try:
                result = await fmi_client.fetch_forecast_harmonie(place)
                rows = parsers.parse_forecast_harmonie(
                    result.data, result.location_metadata, now, place
                )
                count = await forecast_service.upsert_forecasts(db, rows)

                log.status = "success"
                log.records_fetched = count
                log.finished_at = datetime.now(timezone.utc)
                await db.commit()
                logger.info(f"Harmonie forecast for {place}: {count} records")
            except Exception as exc:
                logger.error(f"Harmonie forecast failed for {place}: {exc}")
                log.status = "error"
                log.error_message = str(exc)
                log.finished_at = datetime.now(timezone.utc)
                await db.commit()
